process-engine/process_engine_runtime
# ✨ Auto start on signals and messages

**Changes:**

1. Add `AutoStartService`.
    - Subscribes to new global `signal` and `message` events 
    - Upon receiving an event, it creates new ProcessInstances for each ProcessModel with at least one matching StartEvent.
2. Refactor event based StartEvents:
    - StartEvents with messages or signals will no longer be `suspended`, but continue straight on.
    - The change was necessary, because otherwise the `AutoStartService` would have to publish additional matching Events to the EventAggregator, after each instance was started.
    - As a consequence, ProcessModels with event based StartEvents can now be started regularily as well, which is useful for testing purposes.
3. Message- and Signal- Events will now also publish a global notification that the `AutoStartService` can subscribe to. 
4. Use `^` for devDependencies.
5. Refactor tests for event-based StartEvents and Interprocess-Communication to use the new `AutoStartService`.
6. Adds the columns`state` and `error` to the `Correlation` Data Model. 
7. Apply some refactoring for the `ProcessEngineCore`, which simplifies correlation handling and boosts the performance of the corresponding queries.
8. Reduce the log output generated by the `EventParser` of the `ProcessEngineCore` to reduce log-spamming.
9. Remove the obsolete `@essential-projects/services` packages, along with its contracts, to solve some vulnerabilites that came up recently.
    - All components that previously used that service now use the `HttpClient` directly.
    - The client expects a config located at `config/<environment>/http/http_client`
    - If you have a custom configuration, you can remove the `services` folder completely.

**Issues:**

Closes #229
Closes #231
Closes #246 

PR: #243 (included here)
PR: #247

## How can others test the changes?

- Start the Runtime
- Deploy some ProcessModels that use SignalStartEvents or MessageStartEvents.
    - Best use some suspendable task within those, otherwise this will be very difficult to observe in real-time.
- Send a message or a signal through the ManagementAPI or ConsumerAPI.
- Watch how each ProcessModel with a matching StartEvent gets automatically executed .

## PR-Checklist

Please check the boxes in this list after submitting your PR:
